jupyterhub/apihandlers/users.py

```diff
@@ -183,87 +183,49 @@ class UserServerAPIHandler(APIHandler):
     """
     @gen.coroutine
     @admin_or_self
-    def post(self, name):
+    def post(self, name, server_name=''):
         user = self.find_user(name)
-        if user.running:
+        if server_name:
+            if not self.allow_named_servers:
+                raise web.HTTPError(400, "Named servers are not enabled.")
+        spawner = user.spawners[server_name]
+        if spawner.ready:
             # include notify, so that a server that died is noticed immediately
-            state = yield user.spawner.poll_and_notify()
+            state = yield spawner.poll_and_notify()
             if state is None:
-                raise web.HTTPError(400, "%s's server is already running" % name)
+                raise web.HTTPError(400, "%s's server %s is already running" % (name, server_name))
 
         options = self.get_json_body()
-        yield self.spawn_single_user(user, options=options)
-        status = 202 if user.spawner._spawn_pending else 201
-        if status == 202:
-            self.write(json.dumps({"message": "Server is slow to start"}))
+        yield self.spawn_single_user(user, server_name, options=options)
+        status = 202 if spawner.pending in {'spawn', 'proxy'} else 201
```

Comment:
May want to comment this line re: dictionary of the two pending states 'spawn' and 'proxy'. I'm not quite sure where the 'proxy' state is set.